File: experimental/tools/ndwirerer/get_states.py
def get_state_set(file):
  with open(file) as f:
    text = f.readlines()

  current_state = 0
  states = {}
  for line in text:
    line = line.strip()
    if "-------" in line:
      if current_state > 0:
        states[current_state] = "\n".join(states[current_state])
      current_state += 1
      states[current_state] = []
      continue
    if current_state == 0:
      continue
    if not line.startswith("miter."):
      continue
    states[current_state].append(line)

  del states[current_state]
  return set(states.values())

# TODO cache inf state, maybe needs to be class to be clean
def get_states(inf_file, fin_file):
  
  set_fin = get_state_set(fin_file)

  set_inf = get_state_set(inf_file)

  diff = set_inf - set_fin

  return len(diff)

File: experimental/tools/elastic-miter-generator/GetStates.cpp
#include "llvm/Support/raw_ostream.h"
#include <algorithm>
#include <fstream>
#include <llvm/ADT/StringSet.h>
#include <string>
#include <vector>

namespace dynamatic::experimental {

// TODO handle too many states to print
std::vector<std::string> getStateSet(const std::string &filename) {
  std::ifstream file(filename);
  std::vector<std::string> states;
  std::string line, currentState;
  bool recording = false;

  while (std::getline(file, line)) {

    // TODO put in function
    size_t start = line.find_first_not_of(" \t\n\r\f\v");
    if (start == std::string::npos) {
      line.clear(); // The string contains only whitespace
    } else {
      // Trim leading and trailing spaces
      size_t end = line.find_last_not_of(" \t\n\r\f\v");
      line = line.substr(start, end - start + 1);
    }

    if (line.find("-------") != std::string::npos) {
      if (!currentState.empty()) {
        states.push_back(currentState);
      }
      currentState.clear();
      recording = true;
      continue;
    }

    if (!recording)
      continue;
    // Skip if it doesn't start with "miter."
    if (line.rfind("miter.", 0) != 0) {
      continue;
    }
    currentState += line + "\n";
  }

  if (!currentState.empty()) {
    states.push_back(currentState);
  }

  return states;
}

int getStates(const std::string &infFile, const std::string &finFile) {
  std::vector<std::string> finVector = getStateSet(finFile);
  std::vector<std::string> infVector = getStateSet(infFile);

  // TODO use StringSet directly
  llvm::StringSet<> setFin;
  llvm::StringSet<> setInf;
  for (const auto &entry : finVector) {
    setFin.insert(entry);
  }
  for (const auto &entry : infVector) {
    setInf.insert(entry);
  }

  int diffCount = 0;
  for (const auto &entry : infVector) {
    if (std::find(finVector.begin(), finVector.end(), entry) ==
        finVector.end()) {
      diffCount++;
    }
  }
  return diffCount;
}
} // namespace dynamatic::experimental
